### ISSUE TITLE
GermanCloud FQDN gets wrong suffix - certs/kubeconfig are incorrect

### ISSUE THREAD

**Issue by colemickens** [Azure/acs-engine#450]

**Comment by colemickens** [comment]
For whatever reason, the certs have all the public cloud locations, the china cloud locations, but not the german cloud locations.

**Comment by colemickens** [comment]
```
Unable to connect to the server: dial tcp: lookup kubetest.germanycentral.cloudapp.azure.com on 168.63.129.16:53: no such host
```

**Comment by colemickens** [comment]
cc: @raulfiru

**Comment by raulfiru** [comment]
Hi

Ay progress on this issue? It is the one that's currently preventing us from properly rolling it out on Azure Germany.

**Comment by benkoller** [comment]
The hostname generated by azure (`xxx.germanycentral.cloudapp.microsoftazure.de`) differs from the hostname written to the kubernetes-cfg (`xxx.germanycentral.cloudapp.azure.com`). Changing the hostname in `~/.kube/config` unfortunately did not help much and produced shitloads of these:

```
k8s-master-12345678-0 docker[6474]: E0410 09:04:47.439207    6665 reflector.go:188] pkg/kubelet/kubelet.go:386: Failed to list *api.Node: Get https://y.y.y.y:443/api/v1/nodes?fieldSelector=metadata.name%3Dk8s-master-12345678-0&resourceVersion=0: dial tcp y.y.y.y:443: getsockopt: connection refused
```

Root cause seems to be the usage of the wrong API endpoints as the national and gov clouds use their own flavour:

```
Apr 10 09:06:15 k8s-master-12345678-0 dockerd[6314]: E0410 09:06:15.797716    6665 kubelet_node_status.go:70] Unable to construct api.Node object for kubelet: failed to get external ID from cloud provider: autorest#WithErrorUnlessStatusCode: GET https://login.microsoftonline.de/x-x-x-x/OAuth2/authorize?client_id=x-x-x-x-x&resource=https%3a%2f%2fmanagement.core.windows.net%2f&nc_proxy_request=1&nc_proxy_state=<<someverylongstring>>&nc_client_reply_address=&redirect_uri=https%3a%2f%2flogin.microsoftonline.com%x-x-x-x-x%2fNationalCloudFederation&response_mode=form_post failed with 401 Unauthorized: StatusCode=401
```

Hope this helps.

**Comment by colemickens** [comment]
I just spent time digging into this. @anhowe we really, really need some time to cleanup some stuff in ACS-Engine. We need to pull some of this China/Germany stuff out into a more appropriate location. There are branches buried deep down in code. For just tracing this cert being wrong, I had to go to engine.go, then to acs-engine.go, then had to find out that the thing that formats the FQDN is somehow baked inside the azureconst.go (why?!), and that now inside there, we have even further branching.

We need a proper struct that contains all of the things for a given cloud - the download prefix, the FQDN suffix, etc... and then we need to use a const instance of the struct to pass around everywhere to let specific decisions be made. The `go-autorest` project even has these things already for us to use. For example: https://github.com/Azure/go-autorest/blob/master/autorest/azure/environments.go#L122

**Comment by raulfiru** [comment]
Not clear if you fixed it or it's still pending... but if you have a branch with this fix i would like to test it asap.

Tnx

**Comment by colemickens** [comment]
No, I haven't had time. I think @wangtt03 is going to address this.

**Comment by wangtt03** [comment]
I will handle this issue. Hope I can finish it soon:)

**Comment by wangtt03** [comment]
@colemickens @raulfiru This PR #499 should fix the problem. Could you please verify it on GermanCloud? Thanks!

**Comment by ITler** [comment]
I need to go for AzureGermanCloud *right now*. I use terraform to initially configure Resource Group, vnet, subnet and whatever else is needed, then run `acs-engine generate my_config.json`. All went well with acs-engine 0.50.0 (no programmatic errors raised), but having problems with cube config and certs, that's why I stumbled over this issue here.

To verify PR #499 to work, I cloned master and `make build` my custom executable, manually.

Now I got following error, when`terraform apply` triggers `acs-engine generate my_config.json`:

 	Error running command 'acs-engine generate k8s_rendered.json': exit status 1. Output: time="2017-09-05T17:19:54+02:00" level=info msg="Error returned by LoadContainerService: the service principal client ID must be specified with Orchestrator Kubernetes. Attempting to load container service using LoadContainerServiceForAgentPoolOnlyCluster"
	time="2017-09-05T17:19:54+02:00" level=fatal msg="error parsing the api model: Namespace Properties.DNSPrefix is not caught, Key: 'Properties.DNSPrefix' Error:Field validation for 'DNSPrefix' failed on the 'required' tag\nKey: 'Properties.ServicePrincipalProfile.ClientID' Error:Field validation for 'ClientID' failed on the 'required' tag\nKey: 'Properties.ServicePrincipalProfile.Secret' Error:Field validation for 'Secret' failed on the 'required' tag\nKey: 'Properties.CertificateProfile' Error:Field validation for 'CertificateProfile' failed on the 'required' tag"

My rendered configuration is:

    {
	  "apiVersion": "vlabs",
	  "location": "germanycentral",
	  "properties": {
		"orchestratorProfile": {
		  "orchestratorType": "Kubernetes"
		},
		"masterProfile": {
		  "count": 1,
		  "dnsPrefix": "playground",
		  "vmSize": "Standard_A2",
		  "vnetSubnetId": "/subscriptions/<omitted>/resourceGroups/k8sgermanycentral_playground/providers/Microsoft.Network/virtualNetworks/k8sgermanycentral_playground_vnet/subnets/k8sgermanycentral_playground_vnet_subnet",
		  "firstConsecutiveStaticIP": "172.20.10.10"
		},
		"agentPoolProfiles": [
		  {
			"name": "agentpool1",
			"count": 1,
			"vmSize": "Standard_A2",
			"vnetSubnetId": "/subscriptions/<omitted>/resourceGroups/k8sgermanycentral_playground/providers/Microsoft.Network/virtualNetworks/k8sgermanycentral_playground_vnet/subnets/k8sgermanycentral_playground_vnet_subnet",
			"availabilityProfile": "AvailabilitySet"
		  }
		],
		"linuxProfile": {
		  "adminUsername": "test",
		  "ssh": {
			"publicKeys": [
			  {
				"keyData": "ssh-rsa ..."
			  }
			]
		  }
		},
		"servicePrincipalProfile": {
		  "servicePrincipalClientID": "00000000-...",
		  "servicePrincipalClientSecret": "..."
		}
	  }
	}

Did I miss a change to my k8s config?

**Comment by ITler** [comment]
no one working on this?

**Comment by andyzhangx** [comment]
@ITler it was fixed already, someone could close it?

**Comment by ITler** [comment]
I didn't create this issue, so I didn't close. However, for me this works.

**Comment by ITler** [comment]
@andyzhangx I didn't create this issue, so I didn't close. However, for me this works.
@colemickens is this working for you as well? Might close the issue if it does.

**Comment by raulfiru** [comment]
@colemickens as far as i remember we're using it since some time. @Dutzu you confirm?

**Comment by Dutzu** [comment]
Yes, we have a couple of kubernetes clusters deployed on Azure.DE for some time now...

**Comment by colemickens** [comment]
I think this should just be closed out. However, I no longer have such permisisons.

**Comment by andyzhangx** [comment]
@jdumars could you close this issue? thanks.
